src/Microsoft.Azure.SignalR/HubHost/ServiceLifetimeManagerBase.cs

```diff
@@ -267,6 +289,82 @@ public override Task RemoveFromGroupAsync(string connectionId, string groupName,
+            }
+
+            if (IsInvalidArgument(methodName))
+            {
+                throw new ArgumentException(NullOrEmptyStringErrorMessage, nameof(methodName));
+            }
+            // globally distinct invocationId
+            // $"{connectionId}{_callerId}{_clientResults.GetNewInvocation()}";
+            var invocationId = _clientInvocationManager.Caller.GetNewInvocationId(connectionId, _serverGUID);
+            var task = _clientInvocationManager.Caller.AddInvocation<T>(connectionId, invocationId, cancellationToken);
+            try
+            {
+                var message = AppendMessageTracingId(new ClientInvocationMessage(invocationId, connectionId, _callerId, SerializeAllProtocols(methodName, args, invocationId)));
+                await WriteAsync(message);
+            }
+            catch (Exception)
+            {
+                _clientInvocationManager.Caller.TryRemoveInvocation(invocationId, out var _);
+                throw;
+            }
+
+            try
+            {
+                return await task;
+            }
+            catch
+            {
+                throw;
+            }
+        }
+
+        public override async Task SetConnectionResultAsync(string connectionId, CompletionMessage result)
+        {
+            if (IsInvalidArgument(connectionId))
+            {
+                throw new ArgumentException(NullOrEmptyStringErrorMessage, nameof(connectionId));
+            }
+            if (_clientConnectionManager.ClientConnections.TryGetValue(connectionId, out var clientConnectionContext))
+            {
+                // Current server is a route server.
+                // In order to inform original Caller server with the completion result, send a ClientCompletionMessage to service and the service will route ClientCompletionMessage to the original Caller server.
+                if (_clientInvocationManager.Router.TryGetRoutedInvocation(result.InvocationId, out var _))
+                {
+                    var protocol = clientConnectionContext.Protocol;
+                    var message = AppendMessageTracingId(new ClientCompletionMessage(result.InvocationId, connectionId, _callerId, protocol, SerializeCompletionMessage(result, protocol)));
+                    await WriteAsync(message);
+
+                    _clientInvocationManager.Router.TryCompleteResult(connectionId, result);
+                }
+                else
+                // Current server is the original Caller server. Complete the corresponding client invocation locally.
+                {
+                    _clientInvocationManager.Caller.TryCompleteResult(connectionId, result);
+                }
+            }
+        }
+
+        public override bool TryGetReturnType(string invocationId, [NotNullWhen(true)] out Type type)
+        {
+            if (_clientInvocationManager.Router.TryGetRoutedInvocation(invocationId, out var _))
```

Comment:
looks like the value stored in TryGetRoutedInvocation is never used? Is this method really useful?

Reply:
I delete this method and create a new method `IRoutedClientResultsManager.CheckRoutedInvocation(invocationId)`, which simply check whether a invocation is existing and stored in routed invocations.